File: sdks/shrinkix-php/src/Resources/Optimize.php
<?php

namespace Shrinkix\Resources;

use Shrinkix\Transport;

/**
 * Optimize Resource
 */
class Optimize
{
    private Transport $transport;

    public function __construct(Transport $transport)
    {
        $this->transport = $transport;
    }

    public function optimize(
        $file,
        ?array $resize = null,
        ?array $crop = null,
        ?string $format = null,
        ?int $quality = null,
        ?string $metadata = null
    ): array {
        $multipart = [
            [
                'name' => 'image',
                'contents' => is_string($file) ? fopen($file, 'r') : $file,
            ],
        ];

        if ($resize !== null) {
            $multipart[] = [
                'name' => 'resize',
                'contents' => json_encode($resize),
            ];
        }

        if ($crop !== null) {
            $multipart[] = [
                'name' => 'crop',
                'contents' => json_encode($crop),
            ];
        }

        if ($format !== null) {
            $multipart[] = [
                'name' => 'format',
                'contents' => $format,
            ];
        }

        if ($quality !== null) {
            $multipart[] = [
                'name' => 'quality',
                'contents' => (string)$quality,
            ];
        }

        if ($metadata !== null) {
            $multipart[] = [
                'name' => 'metadata',
                'contents' => $metadata,
            ];
        }

        $result = $this->transport->post('/optimize', [
            'multipart' => $multipart,
        ]);

        return [
            'data' => $result['data'],
            'rate_limit' => $result['rate_limit'],
            'request_id' => $result['rate_limit']['request_id'],
        ];
    }
}

File: sdks/shrinkix-ruby/lib/shrinkix/resources/optimize.rb
# frozen_string_literal: true

require 'json'

module Shrinkix
  module Resources
    # Optimize Resource
    class Optimize
      def initialize(transport)
        @transport = transport
      end

      def optimize(file:, resize: nil, crop: nil, format: nil, quality: nil, metadata: nil)
        # Build multipart payload
        payload = { image: Faraday::UploadIO.new(file, 'image/jpeg') }
        
        payload[:resize] = resize.to_json if resize
        payload[:crop] = crop.to_json if crop
        payload[:format] = format if format
        payload[:quality] = quality.to_s if quality
        payload[:metadata] = metadata if metadata

        result = @transport.post('/optimize', body: payload, multipart: true)

        {
          data: result[:data],
          rate_limit: result[:rate_limit],
          request_id: result[:rate_limit][:request_id]
        }
      end
    end
  end
end
